Support Java 9 class file format

Godin opened the pull request:
https://bugs.openjdk.java.net/browse/JDK-8150011

Godin commented:
@marchof could you please review? Note that this is not sufficient to be able to build JaCoCo with compilation into new class files, because maven-shade-plugin also requires update. However this allows JaCoCo to work properly with new class files - I added integration test (instrumentation with agent + report generation on new class files) and tested with local dev build of latest OpenJDK 9.

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/Java9Support.java:
```diff
@@ -0,0 +1,98 @@
+/**
+ * Patching for Java 9 classes, so that ASM can read them.
+ */
+public final class Java9Support {
```
I see several JavaDoc warnings on this class. Even if the code is temporary I would prefer if our code comes without warnings.

Godin replied:
@marchof I suppose that those warnings are generated by Eclipse? I don't see warnings in IntelliJ , even when explicitly executing all built-in inspections for Javadocs. So could you please elaborate what needs to be fixed?

marchof replied:
Basically we require to have JavaDoc on all public Java classes and members.
* Javadoc: Missing comment for public declaration	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 25	Java Problem
* Javadoc: The method readClass(InputStream, boolean) from the type ClassReader is not visible	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 31	Java Problem
* Javadoc: Missing tag for return type	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 33	Java Problem
* Javadoc: Missing tag for parameter is	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 33	Java Problem
* Javadoc: Missing tag for parameter close	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 33	Java Problem
* Javadoc: Missing tag for declared exception IOException	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 34	Java Problem
* Javadoc: Missing comment for public declaration	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 79	Java Problem
* Javadoc: Missing comment for public declaration	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 83	Java Problem
* Javadoc: Missing comment for public declaration	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 87	Java Problem
* Javadoc: Missing comment for public declaration	Java9Support.java	/org.jacoco.core/src/org/jacoco/core/internal	line 94	Java Problem

marchof commented on org.jacoco.core/src/org/jacoco/core/internal/Java9Support.java:
```diff
@@ -0,0 +1,98 @@
+			throw new IOException("Class not found");
+		}
+		try {
+			byte[] b = new byte[is.available()];
```
This code has quite some complexity. We should either add tests or use more high level APIs. E.g. the input can be easily copied into a ByteArrayOutputStream which then most of the copy magic for us (needs a little bit more memory allocations though).

marchof commented:
I wonder whether we should drop `Analyzer.analyzeClass(ClassReader)` from the public API? This is a asymmetry to Instrumenter anyways.

marchof commented:
@Godin Beside my comments I'm ok with this PR!

Godin commented:
@marchof I simplified code and added Javadocs.

About asymmetry: indeed it bothers. This is the only place where we expose ClassReader, but not the only place where we expose ASM APIs. Removal - is an API change. I don't think that it is critical for experimental support of Java 9 class file format. So IMO we should consider separately question about marking it as deprecated and further removal. WDYT?

marchof commented:
@Godin I agree to remove this API in a separate request. I just came across this as it won't work for Java 9.

bjkail commented:
FYI (sorry if this is the wrong forum for this comment), we override analyzeClass(ClassReader) in order to filter which classes should be analyzed: for historical reasons, module boundaries in our product are not cleanly separated by archive, so we use package filters here.  It would be ideal if there were still some way for us to do that efficiently before you remove that method (we can also skip the ICoverageVisitor.visitCoverage event, but that's less efficient since Analyzer's ASM processing still occurs).

Godin commented:
@bjkail No need to apologize. In fact thank you for this valuable information.

As far as I understood you're doing something like
```
Analyzer analyzer = new Analyzer(...);
ClassReader classReader = new ClassReader(...);
if (notFiltered(reader)) {
  analyzer.analyzeClass(reader);
}
```
if so, wondering whether you can replace it by
```
Analyzer analyzer = new Analyzer(...);
ClassReader classReader = new ClassReader(...);
if (notFiltered(reader)) {
  analyzer.analyzeClass(reader.b, "");
}
```
?

bjkail commented:
Well, more like this:

    Analyzer analyzer = new Analyzer(...) {
      @Override
      public void analyzeClass(ClassReader r) {
        if (notFiltered(r.getClassName())) {
          super.analyzeClass(r);
        }
      }
    };
    
    // Use Analyzer's existing recursive archive processing logic.
    analyzer.analyzeAll(...);

Regardless, we could instead override `analyzeClass(InputStream, String)` to do as you suggest at the cost of a duplicate ClassReader constructor call, but I suppose that's relatively lightweight when compared to overriding visitExecution.  Well, the Analyzer javadoc doesn't actually guarantee that analyzeAll will invoke the analyzeClass methods, but it is nice to be able to reuse Analyzer's recursive archive processing, so it would be nice if the javadoc did make that guarantee.

marchof commented:
For the filtering API story we will need proper call-backs anyways.

Godin commented:
JDK 9 EA build 119 is available for testing, so let's go ahead with merging of those changes. And let's handle raised points separately from this PR, if it is needed or would be needed.

And for the record without those changes execution of JaCoCo with JDK 9 EA b 119 fails with message like
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-ea/Native Method)
     at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-ea/NativeMethodAccessorImpl.java:62)
     at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-ea/DelegatingMethodAccessorImpl.java:43)
     at java.lang.reflect.Method.invoke(java.base@9-ea/Method.java:531)
     at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(java.instrument@9-ea/InstrumentationImpl.java:396)
     at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(java.instrument@9-ea/InstrumentationImpl.java:408)
Caused by: java.lang.RuntimeException: Class java/util/UUID could not be instrumented.
     at org.jacoco.agent.rt.internal_9472e99.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:139)
     at org.jacoco.agent.rt.internal_9472e99.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:100)
     at org.jacoco.agent.rt.internal_9472e99.PreMain.createRuntime(PreMain.java:55)
     at org.jacoco.agent.rt.internal_9472e99.PreMain.premain(PreMain.java:47)
     ... 6 more
Caused by: java.lang.NoSuchFieldException: $jacocoAccess
     at java.lang.Class.getField(java.base@9-ea/Class.java:1881)
     at org.jacoco.agent.rt.internal_9472e99.core.runtime.ModifiedSystemClassRuntime.createFor(ModifiedSystemClassRuntime.java:137)
     ... 9 more
```
which could be improved - something like "cannot read class".
